Update rules library version

CurlyMoo opened the pull request:
This should fix issue https://github.com/Egyras/HeishaMon/issues/557 and https://github.com/Egyras/HeishaMon/issues/573

IgorYbema commented:
@CurlyMoo can I ask you to add support for the relays on the large heishamon to be controlled using rules? a new PR for this or in this PR is ok

IgorYbema commented:
https://github.com/IgorYbema/HeishaMon/blob/3018485d16e7ee9777388b1f2b99b9ac7502c78e/HeishaMon/gpio.cpp#L14

stumbaumr commented:
My current rules are no longer working with the current build from https://github.com/IgorYbema/HeishaMon/actions/runs/12723065851

IgorYbema commented:
@stumbaumr  I just loaded this firmware and your rules and it worked fine (on a large heishamon EPS32). Can you tell me what you did different?

stumbaumr commented:
I am using a ESP8266 - the small one...

IgorYbema commented:
It is loading fine on that same build on a ESP8266 heishamon with same settings in my test setup. So don't know why it is failing for you.  Can you paste the logs while loading the rules with this firmware if you want to?

stumbaumr commented:
Here is the log:
```
Sun Jan 12 22:48:39 2025 (157512): Calculated Watt on S0 port 1: 1618
rule #1 was prepared in 5774 microseconds
rule #1 bytecode was created in 13478 microseconds
bytecode: 48/48, heap: 32/32, stack: 4/4 bytes, varstack: 64/64 bytes
timer #10 set to 40 seconds
rule #1 was executed in 3202 microseconds
bytecode: 48/48, heap: 32/32, stack: 4/12 bytes, varstack: 64/64 bytes
rule #2 was prepared in 227215 microseconds
rule #2 bytecode was created in 650317 microseconds
bytecode: 496/496, heap: 92/92, stack: 4/12 bytes, varstack: 184/184 bytes
Sun Jan 12 22:48:43 2025 (160929): #maxPumpDutyBeforeDHW
ERROR: cannot compare < with a left char value
rules memory used: 1152 / 16384
Sun Jan 12 22:48:43 2025 (160941): Failed to load new rules, reverting back to older rules!
```
Tested on this build:
https://github.com/IgorYbema/HeishaMon/actions/runs/12736286709
with these rules:
```
on System#Boot then
   #stateBeforeDHW = 1;
   #OpModeBeforeDHW = 0;
   #targetLowBeforeDHW = 31;
   #targetHighBeforeDHW = 40;
   #maxPumpDutyBeforeDHW = 110;
   #offAtNight = 1;
   @SetPump = 0;
   setTimer(10, 40);
end

on timer=10 then
   setTimer(10, 30);
   $hour = %hour;
   $minute = %minute;
   if $hour == "6" && $minute == "0" then
      if @Outside_Temp < 12 then
         #maxPumpDutyBeforeDHW = 160;
         @SetMaxPumpDuty = #maxPumpDutyBeforeDHW;
      end
      if @Operating_Mode_State == 0 && #offAtNight == 2 && @Heatpump_State == 0 then
         #offAtNight = 1;
         @SetHeatpump = 1;
      end
   end
   if $hour == 9 && $minute == 0 && @Operating_Mode_State == 1 && #offAtNight == 2 && @Heatpump_State == 0 && @Outside_Temp > 27 then
      #offAtNight = 1;
      @SetHeatpump = 1;
   end
   if (( %hour == 13 && %minute == 0 && @DHW_Temp < ( @DHW_Target_Temp + @DHW_Heat_Delta )) || @DHW_Temp < 20 ) && @Operating_Mode_State < 3 then
      #stateBeforeDHW = @Heatpump_State;
      #OpModeBeforeDHW = @Operating_Mode_State;
      @SetHeatpump = 1;
      @SetOperationMode = #OpModeBeforeDHW + 4;
   end
   if $hour == 18 && $minute == 0 && #offAtNight == 1 && @Heatpump_State == 1 then
      if @Operating_Mode_State == 0 && @Room_Thermostat_Temp > 22 && @Outside_Temp > 9 then
         #offAtNight = 2;
      end
      if @Operating_Mode_State == 1 then
         #offAtNight = 2;
         @SetHeatpump = 0;
      end
   end
   if $hour == 20 && $minute == 0 then
      #maxPumpDutyBeforeDHW = 110;
      @SetMaxPumpDuty = #maxPumpDutyBeforeDHW;
      if @Operating_Mode_State == 0 && #offAtNight == 2 && @Heatpump_State == 1 then   
         @SetHeatpump = 0;
      end
   end
end

on @DHW_Power_Consumption then
   if @Heatpump_State == 1 && @Operating_Mode_State >= 3 && @DHW_Power_Consumption == 0 && ( @DHW_Temp >= @DHW_Target_Temp || @Main_Outlet_Temp >= @Main_Target_Temp ) then
      @SetHeatpump = #stateBeforeDHW;
      @SetOperationMode = #OpModeBeforeDHW;
   end
end
on @ThreeWay_Valve_State then
   if @ThreeWay_Valve_State == 1 then
      #targetLowBeforeDHW = @Z1_Heat_Curve_Target_Low_Temp;
      #targetHighBeforeDHW = @Z1_Heat_Curve_Target_High_Temp;
      #maxPumpDutyBeforeDHW = @Max_Pump_Duty;
      @SetOperationMode = #OpModeBeforeDHW + 4;
      $minDHWTarget = 45;
      $minOutside = 10;
      $maxDHWTarget = 52;
      $maxOutside = 25;
      $newDHWTarget = round($minDHWTarget + ((@Outside_Temp - $minOutside ) * ($maxDHWTarget - $minDHWTarget) / ($maxOutside - $minOutside)));
      if @Outside_Temp <= $minOutside then
         $newDHWTarget = $minDHWTarget;
      end
      if @Outside_Temp >= $maxOutside then
         $newDHWTarget = $maxDHWTarget;
      end
      if @DHW_Target_Temp != $newDHWTarget then
         @SetDHWTemp = $newDHWTarget;
      end
      $targetAfterDHW = ( @DHW_Target_Temp + 5 );
      @SetCurves = concat('{zone1:{heat:{target:{high:', $targetAfterDHW ,',low:', $targetAfterDHW ,'}}}}');
      @SetMaxPumpDuty = 160;
      @SetPump = 1;	  
   else
      @SetHeatpump = #stateBeforeDHW;
      @SetOperationMode = #OpModeBeforeDHW;
      @SetPump = 1;
      setTimer(20, 240);
   end
end
on timer=20 then
   @SetZ1HeatRequestTemperature = 5;
   @SetCurves = concat('{zone1:{heat:{target:{high:', #targetHighBeforeDHW ,',low:', #targetLowBeforeDHW ,'}}}}');
   @SetMaxPumpDuty = #maxPumpDutyBeforeDHW;
   @SetPump = 0;
end

on @Main_Target_Temp then
   throttle();
end
on @Main_Outlet_Temp then
   throttle();
end
on @Main_Inlet_Temp then
   throttle();
end
on throttle then
   if @Heatpump_State == 1 then
      if @Operating_Mode_State % 2 == 0 || @ThreeWay_Valve_State == 1 then
         #roomDelta = 0;
         if @Outside_Temp < 7 && @Room_Thermostat_Temp > 22 then
            #roomDelta = -1;
         end
         #outletDeltaHeat = @Main_Outlet_Temp  - @Main_Target_Temp + @Z1_Heat_Request_Temp - #roomDelta;
         if #outletDeltaHeat < 0.3 then
            $newQuietMode = 0;
         end
         if #outletDeltaHeat >= 1.6 then
            $newQuietMode = 3;
         end
         #newHeatRequest = floor( #outletDeltaHeat ) - 1;
         if #newHeatRequest > 5 then
            #newHeatRequest = 5;
            if #offAtNight == 1 && @Outside_Temp > 9 then
               #offAtNight = 2;
            end
         end
         if #newHeatRequest <= 0 && ( @Main_Outlet_Temp  - @Main_Target_Temp ) < 2 then
            #newHeatRequest = #roomDelta;
         end
         if @Z1_Heat_Request_Temp != #newHeatRequest then
            @SetZ1HeatRequestTemperature = #newHeatRequest;
         end
      else
         $outletDeltaCool = @Main_Outlet_Temp - @Main_Target_Temp + @Z1_Cool_Request_Temp;
         if $outletDeltaCool > 0 then
            $newQuietMode = 0;
         end
         if $outletDeltaCool <= -1 then
            $newQuietMode = 3;
         end
         $newCoolRequest = floor( $outletDeltaCool ) + 1;
         if $newCoolRequest < -2 then
            $newCoolRequest = -2;
         end
         if $newCoolRequest > 3 then
            $newCoolRequest = 3;
         end
         if @Z1_Cool_Request_Temp != $newCoolRequest then
            @SetZ1CoolRequestTemperature = $newCoolRequest;
         end
      end
      if @Quiet_Mode_Level != $newQuietMode then
         @SetQuietMode = $newQuietMode;
      end
   end
end
```

stumbaumr commented:
And I am back to https://github.com/IgorYbema/HeishaMon/actions/runs/12714977590

CurlyMoo commented:
@IgorYbema What's the onboot default value of `@Outside_Temp`? Could it be that that value triggers the push of a string value? 
https://github.com/IgorYbema/HeishaMon/blob/3.9b/HeishaMon/rules.cpp#L442-L458

Maybe the `check_is_number` contains a bug?
https://github.com/IgorYbema/HeishaMon/blob/3.9b/HeishaMon/rules.cpp#L351-L376

IgorYbema commented:
Hmm it doesn't fail when the outside_temp is default (-128) but it fails when it has a valid value.
That is why I could not reproduce the error as I tried it on my test setup which doesn't have active data. It does fail on my heishamon which is connected to the heatpump

CurlyMoo commented:
>Why do you not use strtod() instead of creating your own check_is_number? Is that not in stdlib.h?

Maybe we can try that instead? Can you do that first in your local context?

IgorYbema commented:
This is wrong. Should be < and not <= I believe
`while(pos <= len) {`

CurlyMoo commented:
I have a similar function inside the rules library to parse number from a stream of data. Quit possible i made a mistake in their.

> Why do you not use strtod() instead of creating your own check_is_number?

It's also always a good idea to give this feedback when the PR is still open 😉

IgorYbema commented:
@stumbaumr  you can test this build when the actions is finished building it: https://github.com/IgorYbema/HeishaMon/actions/runs/12772225974

I will test it myself also after dinner

CurlyMoo commented:
This was the suggestion i got:

> The function provided attempts to check if a given string represents a number by iterating through each character and applying certain conditions. However, there are a few issues with the logic and implementation that need to be addressed to make it work correctly. Here’s a corrected version of the function with explanations:

```c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <ctype.h>

static int8_t check_is_number(char *str) {
    uint16_t i = 0, nrdot = 0, pos = 0, len = strlen(str);
    char current;

    // Check if the string is empty
    if (len == 0) {
        return -1;
    }

    current = str[0];

    // Check the first character
    if (isdigit(current) || current == '-') {
        /* 
         * The dot cannot be the first character 
         * and we cannot have more than 1 dot 
         */
        while (pos < len) {
            current = str[pos];

            if (!(isdigit(current) || (i == 0 && current == '-') || (i > 0 && nrdot == 0 && current == '.'))) {
                return -1; // Invalid character found
            }

            if (current == '.') {
                nrdot++;
            }

            pos++;
            i++;
        }

        return 0; // It is a valid number
    } else {
        return -1; // Doesn't start with a digit or '-'
    }
}
```

# Changes Made:
## Boundary Condition Check:

* Added a check `if (len == 0)` at the beginning to handle empty strings. An empty string should not be considered a number.

## Loop Condition:

* Changed `while (pos <= len)` to `while (pos < len)`. This ensures that the loop terminates before attempting to access `str[len]`, which is out of bounds.

## Character Fetching:

* Moved `current = str[pos]` inside the `while` loop to correctly fetch the current character on each iteration.

## Return Values:

Changed `return 0`; inside the loop to `return 0`; after the loop to signify that all characters were valid.

# Logic Corrections:

## Dot Counting:

* Counting of dots (`.`) is correctly handled with `nrdot` to ensure there's no more than one dot and that it cannot be the first character.

## First Character Handling:

* The first character can be a digit (`0-9`) or a minus sign (`-`). This is checked explicitly.

## Subsequent Characters:

* For subsequent characters, digits (`0-9`), a minus sign (only if it's the first character), and a dot (only if it hasn't been encountered before) are allowed.

# Additional Considerations:

* This function assumes ASCII characters and does not handle floating-point exponent notation (`e` or `E`). If you need to support scientific notation, additional logic would be necessary.
* Make sure to include necessary header files (`<stdio.h>`, `<stdint.h>`, `<string.h>`, `<ctype.h>`) for the function to compile and work correctly in a C environment.

By applying these corrections, the function should now correctly determine whether the input string represents a valid number based on the specified criteria.

stumbaumr commented:
Excellent work, ChatGPT!
It is working for me now and reading that code part became easier as well!
Running on https://github.com/IgorYbema/HeishaMon/actions/runs/12775259241 now...

CurlyMoo commented:
But before someone complains why there is a need for this function:
https://github.com/IgorYbema/HeishaMon/blob/3.9b/HeishaMon/src/rules/rules.cpp#L298-L326

That's because this functions needs to be able to handle byte-by-byte memory reading as required by the 2nd heap. `stdtod` doesn't do that safe reading internally.

CurlyMoo commented:
> Isn't that function wrong also as it also does pos <= len . Did you check?

https://github.com/CurlyMoo/rules/commit/6cd22379fb1a177000fa487dc10d468e9fb8d238

IgorYbema commented:
> Yes I confirm this 2nd bug. I try to look which can cause this. #targetLowBeforeDHW is never changed in the timer routine but get changed anyway.

Ok I just confirmed that when reverting this PR that this issue gone. I'll try to revert the commits one by one of the PR to see where it went wrong

IgorYbema commented:
> Yes I confirm this 2nd bug. I try to look which can cause this. #targetLowBeforeDHW is never changed in the timer routine but get changed anyway.

I just reverted this commit for this PR and opened a new PR so we can handle the rules library update from there properly

https://github.com/IgorYbema/HeishaMon/pull/153
